File: sdks/node/src/crypto.js
const { spawnSync } = require('child_process');
const path = require('path');
const fs = require('fs');

function findHelper() {
  if (process.env.NIVRIT_CRYPTO_HELPER) return process.env.NIVRIT_CRYPTO_HELPER;
  const exe = process.platform === 'win32' ? 'nivrit-crypto-helper.exe' : 'nivrit-crypto-helper';
  // Bundled platform package, installed via optionalDependencies (the package
  // manager picks the one matching os+cpu). See scripts/build-platform-packages.sh.
  try {
    return require.resolve(`@nivrit/helper-${process.platform}-${process.arch}/bin/${exe}`);
  } catch (_) { /* not installed for this platform — fall through */ }
  // Dev fallback: monorepo cargo build output.
  for (const sub of ['release', 'debug']) {
    const c = path.join(__dirname, '..', '..', '..', 'target', sub, exe);
    if (fs.existsSync(c)) return c;
  }
  throw new Error(
    'nivrit-crypto-helper not found. Reinstall @nivrit/sdk on a supported platform, ' +
    'set NIVRIT_CRYPTO_HELPER, or run `cargo build --release -p nivrit-crypto-helper`.'
  );
}

class HelperCrypto {
  constructor(helperPath) {
    this.helperPath = helperPath || findHelper();
  }

  call(req) {
    const result = spawnSync(this.helperPath, [], {
      input: JSON.stringify(req),
      encoding: 'utf-8',
      timeout: 30000,
    });
    if (result.error) throw new Error(`crypto helper failed: ${result.error.message}`);
    if (result.status !== 0) throw new Error(`crypto helper exited ${result.status}: ${result.stderr}`);
    let resp;
    try {
      resp = JSON.parse(result.stdout.trim());
    } catch (e) {
      throw new Error(`crypto helper returned invalid JSON: ${result.stdout}`);
    }
    if (!resp.ok) throw new Error(resp.error || 'crypto helper error');
    return resp.result;
  }

  hybridSuiteId() {
    return this.call({ op: 'hybrid_suite_id' });
  }

  generateKeypair(password) {
    return this.call({ op: 'generate_keypair', password });
  }

  decryptPrivateKey(encryptedPrivateKey, nonce, password) {
    return this.call({
      op: 'decrypt_private_key',
      encrypted_private_key: encryptedPrivateKey,
      nonce,
      password,
    }).private_key;
  }

  decapsulateProjectKey(encryptedProjectKey, privateKey) {
    return this.call({
      op: 'decapsulate_project_key',
      encrypted_project_key: encryptedProjectKey,
      private_key: privateKey,
    }).project_key;
  }

  encryptValue(plaintext, key) {
    return this.call({ op: 'encrypt_value', plaintext, key });
  }

  decryptValue(ciphertext, nonce, key) {
    return this.call({ op: 'decrypt_value', ciphertext, nonce, key }).plaintext;
  }

  encapsulateProjectKey(projectKey, recipientPublicKey) {
    return this.call({
      op: 'encapsulate_project_key',
      project_key: projectKey,
      recipient_public_key: recipientPublicKey,
    });
  }
}

module.exports = { HelperCrypto };

File: sdks/ruby/lib/nivrit_sdk/crypto.rb
require 'json'
require 'open3'
require 'pathname'

module NivritSdk
  class HelperCrypto
    def initialize(helper_path = nil)
      @helper_path = helper_path || find_helper
    end

    def hybrid_suite_id
      call('op' => 'hybrid_suite_id')
    end

    def generate_keypair(password)
      call('op' => 'generate_keypair', 'password' => password)
    end

    def decrypt_private_key(encrypted_private_key, nonce, password)
      call(
        'op' => 'decrypt_private_key',
        'encrypted_private_key' => encrypted_private_key,
        'nonce' => nonce,
        'password' => password
      )['private_key']
    end

    def decapsulate_project_key(encrypted_project_key, private_key)
      call(
        'op' => 'decapsulate_project_key',
        'encrypted_project_key' => encrypted_project_key,
        'private_key' => private_key
      )['project_key']
    end

    def encrypt_value(plaintext, key)
      call('op' => 'encrypt_value', 'plaintext' => plaintext, 'key' => key)
    end

    def decrypt_value(ciphertext, nonce, key)
      call(
        'op' => 'decrypt_value',
        'ciphertext' => ciphertext,
        'nonce' => nonce,
        'key' => key
      )['plaintext']
    end

    def encapsulate_project_key(project_key, recipient_public_key)
      call(
        'op' => 'encapsulate_project_key',
        'project_key' => project_key,
        'recipient_public_key' => recipient_public_key
      )
    end

    private

    def find_helper
      env = ENV['NIVRIT_CRYPTO_HELPER']
      return env if env && !env.empty?

      exe = Gem.win_platform? ? 'nivrit-crypto-helper.exe' : 'nivrit-crypto-helper'

      # Bundled in the platform gem at sdks/ruby/vendor/ (see build-gems.sh).
      # __dir__ is sdks/ruby/lib/nivrit_sdk, so vendor/ is two levels up.
      bundled = File.expand_path(File.join(__dir__, '..', '..', 'vendor', exe))
      return bundled if File.exist?(bundled)

      # Dev fallback: monorepo cargo build output (repo root is four levels up).
      here = Pathname.new(__dir__).expand_path
      (here / '..' / '..' / '..' / '..' / 'target' / 'release' / exe).to_s
    end

    def call(req)
      out, status = Open3.capture2(@helper_path, stdin_data: req.to_json)
      raise "crypto helper exited #{status.exitstatus}" unless status.success?

      resp = JSON.parse(out.strip)
      raise "crypto helper error: #{resp['error']}" unless resp['ok']

      resp['result']
    end
  end
end
